src/Renderer/LayeredMaterial.js

```diff
@@ -67,459 +132,233 @@ export function getMaxColorSamplerUnitsCount() {
-    let offsetLayer = 0;
-    let offsetTexture = 0;
-
-    const originalOffsets = new Array(...this.uniforms.offsetScale_L01.value);
-    const originalTextures = new Array(...this.uniforms.dTextures_01.value);
-
-    for (let l = 0; l < sequenceLayer.length; l++) {
-        const layer = sequenceLayer[l];
-        const oldIndex = this.indexOfColorLayer(layer);
-        if (oldIndex > -1) {
-            const newIndex = l - offsetLayer;
-            const texturesCount = this.layerTexturesCount[oldIndex];
-
-            // individual values are swapped in place
-            if (newIndex !== oldIndex) {
-                moveElementArray(this.colorLayersId, oldIndex, newIndex);
-                moveElementArray(this.layerTexturesCount, oldIndex, newIndex);
-                moveElementArray(this.uniforms.paramLayers.value, oldIndex, newIndex);
-                moveElementArray(this.uniforms.visibility.value, oldIndex, newIndex);
-            }
-            const oldOffset = this.getTextureOffsetByLayerIndex(newIndex);
-            // consecutive values are copied from original
-            for (let i = 0; i < texturesCount; i++) {
-                this.uniforms.offsetScale_L01.value[offsetTexture + i] = originalOffsets[oldOffset + i];
-                this.uniforms.dTextures_01.value[offsetTexture + i] = originalTextures[oldOffset + i];
+function setUniformProperty(object, property, initValue) {
+    object.uniforms[property] = new THREE.Uniform(initValue);
+    Object.defineProperty(object, property, {
+        get: () => object.uniforms[property].value,
+        set: (value) => {
+            if (object.uniforms[property].value != value) {
+                object.uniforms[property].value = value;
+                object.uniformsNeedUpdate = true;
             }
+        },
+    });
+}
 
-
-            this.setTextureOffsetByLayerIndex(newIndex, offsetTexture);
-            offsetTexture += texturesCount;
-        } else {
-            offsetLayer++;
-        }
-    }
-
-    this.uniforms.colorLayersCount.value = this.getColorLayersCount();
+const ELEVATION_MODES = {
+    RGBA: 0,
+    COLOR: 1,
+    DATA: 2,
 };
 
-LayeredMaterial.prototype.removeColorLayer = function removeColorLayer(layer) {
-    const layerIndex = this.indexOfColorLayer(layer);
+class LayeredMaterial extends THREE.RawShaderMaterial {
+    constructor(options = {}) {
+        super({});
 
-    if (layerIndex === -1) {
-        return;
-    }
-
-    const offset = this.getTextureOffsetByLayerIndex(layerIndex);
-    const texturesCount = this.getTextureCountByLayerIndex(layerIndex);
-
-    // remove layer
-    this.colorLayersId.splice(layerIndex, 1);
-    this.uniforms.colorLayersCount.value = this.getColorLayersCount();
-
-    // remove nb textures
-    this.layerTexturesCount.splice(layerIndex, 1);
-    this.layerTexturesCount.push(0);
+        const nbSamplers = [samplersElevationCount, getMaxColorSamplerUnitsCount()];
```

Comment:
I don't really like this part: the higher `getMaxColorSamplerUnitsCount` gets, the higher the number of initialized uniform there is. And if uniforms are initialized, they will be called even if their value is `undefined`. So that means that we will have a high number of draw calls for nothing (here I get from ~1000 DC to ~2800DC for a single frame, with most DC contributing nothing). I get that this is supposed to ease things and how we use them, but can't we make a change here ?